File: scala/src/main/scala/leetcode/editor/cn/[3]无重复字符的最长子串.scala
package leetcode.editor.cn
object LongestSubstringWithoutRepeatingCharacters {
//leetcode submit region begin(Prohibit modification and deletion)
object Solution {
  import scala.collection.mutable
    def lengthOfLongestSubstring(s: String): Int = {
      val window = mutable.Map.empty[Char,Int]
      var left = 0
      var right =0
      var res = 0

      while (right < s.length){
        val c = s(right)
        right += 1
        window.update(c,window.getOrElse(c,0)+1)

        while (window(c) > 1) {
         val d = s(left)
          left += 1
          window.updateWith(d){
            case Some(v) if v > 1 => Some(v-1)
            case _ => None
          }
        }

        res = math.max(res,right - left)
      }

      res
    }
}
//leetcode submit region end(Prohibit modification and deletion)

}

File: go/leetcode/editor/cn/[3]无重复字符的最长子串.go
package leetcode

// leetcode submit region begin(Prohibit modification and deletion)
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func lengthOfLongestSubstring(s string) int {
	var window = make(map[byte]int)
	var left = 0
	var right = 0
	var res = 0

	for right < len(s) {
		var c = s[right]
		right++
		window[c]++
		for window[c] > 1 {
			var d = s[left]
			left++
			window[d]--
		}
		res = max(res, right-left)
	}
	return res
}

//leetcode submit region end(Prohibit modification and deletion)
